### PR TITLE
Legacy multisite compatibility

### LINKED ISSUE #33: Uploading to bucket root instead of wp-content/blogs.dir/1/files/
Hi, thanks for letting us use s3 to host our uploaded files!

When trying out this plugin in our multisite environment, I can't find out how to prefix uploaded files with the configured upload path per site, which means the different sites may overwrite each other's files.

When a site's upload path is configured as `wp-content/blogs.dir/1/files/` and I upload `somefile.ext`, I want it to end up like `<bucket>/blogs.dir/1/files/somefile.ext`. Instead it ends up like `<bucket>/somefile.ext`.

Is there a simple configuration to enable this that I've overlooked?

### LINKED ISSUE #200: Upgraded legacy Multisite URLs do not have subsite number in them
If a MS install created pre 3.5 has been upgraded to 3.5+, the base upload path derived from the option 'upload_path' still contains the the full site path for files, e.g `wp-content/blogs.dir/2/files`. This means that when it is removed from the `wp_upload_dir` path it will remove the subsite identifier, resulting in a generic prefix across all the sites in the network.

### PR DESCRIPTION
**Pull request by polevaultweb**
Support subsite paths for MS that were created pre 3.5

Resolves #200 
Resolves #33
Resolves #175

### CODE REVIEW

**Comment by bradt** [comment]
If my files are in `wp-content/blogs.dir/2/files` on my server, it will store them in `wp-content/sites/2/`? Shouldn't we maintain the same directory structure as on the server?

**Comment by polevaultweb** [comment]
> Shouldn't we maintain the same directory structure as on the server?
We don't in other places, and let users change the prefix so it can be wildly different.

Tbh, having `/blogs.dir/` in a path/URL i thought looked ugly

**Comment by bradt** [comment]
Well, the current behaviour is to mirror what the current filesystem does. And it's what we say in the UI:

image

**Comment by polevaultweb** [comment]
@bradt ok cool, reverted it to what is is with a filter

**Comment by bradt** [comment]
Actually, looks like I was wrong about `blogs.dir` in the URL...

> This is a change from Multisite 3.0-3.4.2, where images of subsites were stored in /wp-content/blogs.dir/ and were shown in http://example.com/files/ and http://example.com/sitename/files and so on. If you started with a Multisite install older than 3.5, it is not an error if your images show with the URL of /files/.

http://codex.wordpress.org/Multisite_Network_Administration#Uploaded_File_Path

Can we mimic this default? I'm not sure what "sitename" is exactly.

File: classes/amazon-s3-and-cloudfront.php [hunk 1]
```diff
@@ -1326,6 +1326,14 @@ function get_dynamic_prefix( $time = null ) {
...
 		}
 
+		// support legacy MS installs (<3.5 since upgraded) for subsites
+		if ( is_multisite() && 1 != ( $blog_id = get_current_blog_id() ) && strpos( $prefix, 'sites/' ) === false ) {
+			$legacy_ms_prefix = '/blogs.dir/' . $blog_id . '/files/';
+			$legacy_ms_prefix = apply_filters( 'as3cf_legacy_ms_subsite_prefix', $legacy_ms_prefix, $blog_id );
+
+			$prefix = '/' . trailingslashit( ltrim( $legacy_ms_prefix, '/' ) ) . ltrim( $prefix, '/' );;
+		}
+
```
Comment (bradt) [inline classes/amazon-s3-and-cloudfront.php]: Won't this result in URLs of the form `/2015/01/blogs.dir/2/files/` ? Seems backward to me.

**Comment by polevaultweb** [comment]
`sitename` is when the network is a sub directory install. But that won't work for sudomain installs or domain mapping subsites.

File: classes/amazon-s3-and-cloudfront.php (hunk 1 above)
Comment (polevaultweb) [inline classes/amazon-s3-and-cloudfront.php]: No, it will be `/blogs.dir/2/files/2015/01`

**Comment by bradt** [comment]
Ah I see. So why won't it work for a subdomain install? Isn't `sitename` then just at the beginning of the URL? i.e. http://sitename.example.com/files

If so, it should be fine for us to use `sitename` in our URLs to mimic the default behaviour as closely as possible. e.g. http://bucketname.s3.amazonaws.com/sitename/files/2015/01/photo.jpg

**Comment by polevaultweb** [comment]
Yep, good call

File: classes/amazon-s3-and-cloudfront.php (hunk 1 above)
Comment (bradt) [inline classes/amazon-s3-and-cloudfront.php]: Yep, see it now.

**Comment by bradt** [comment]
Setup a 3.4.2 install, then upgraded to 4.1 to test this. All good.

**Pull request merged**